### ISSUE TITLE
Improve get_state functionality 

### ISSUE THREAD

**Issue by zohimchandani** [NVIDIA/cuda-quantum#822]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

```

import cudaq 

kernel = cudaq.make_kernel()

qubits = kernel.qalloc(2)

kernel.x(qubits[0])
kernel.x(qubits[0])

state = cudaq.get_state(kernel)

state.dump()
```

```
1+0j 0+0j 0+0j 0+0j 
```

The code above dumps the state but the output can't be manipulated as it does not have a `type`. I suggest it be dictionary where the keys are the basis states and the values are the complex amplitudes. 

Moreover, it is not clear that the basis state ordering starts from the 11 state. A dictonary key value mapping would solve that too. 

### Steps to reproduce the bug

NA

### Expected behavior

NA

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

-

**Comment by zohimchandani** [comment]
The issue of manipulating the coefficients can be solved by this:

```
import cudaq 
import numpy as np 

kernel = cudaq.make_kernel()

qubits = kernel.qalloc(2)

kernel.x(qubits[0])
kernel.x(qubits[0])

state = cudaq.get_state(kernel)

npa = np.array(state)

```

however the labelling of the basis states still remains and I suggest a dictonary style structure to represent this information.

**Comment by bettinaheim** [comment]
addressed by #1107
